src/protocol_common.c

```diff
@@ -114,6 +122,132 @@ esp_loader_error_t loader_flash_end_cmd(bool stay_in_loader)
+            0x55, 0x55, 0x55, 0x55, 0x55, 0x55, 0x55, 0x55,
+            0x55, 0x55, 0x55, 0x55, 0x55, 0x55, 0x55, 0x55,
+            0x55, 0x55, 0x55, 0x55, 0x55, 0x55, 0x55, 0x55,
+        }
+    };
+
+    return send_cmd(&sync_cmd, sizeof(sync_cmd), NULL);
+}
+
+
+esp_loader_error_t loader_spi_attach_cmd(uint32_t config)
+{
+    spi_attach_command_t attach_cmd = {
+        .common = {
+            .direction = WRITE_DIRECTION,
+            .command = SPI_ATTACH,
+            .size = CMD_SIZE(attach_cmd),
+            .checksum = 0
+        },
+        .configuration = config,
+        .zero = 0
+    };
+
+    return send_cmd(&attach_cmd, esp_stub_running ? sizeof(attach_cmd) - sizeof(attach_cmd.zero) : sizeof(attach_cmd), NULL);
+}
+
+
+esp_loader_error_t loader_md5_cmd(uint32_t address, uint32_t size, uint8_t *md5_out)
+{
+    spi_flash_md5_command_t md5_cmd = {
+        .common = {
+            .direction = WRITE_DIRECTION,
+            .command = SPI_FLASH_MD5,
+            .size = CMD_SIZE(md5_cmd),
+            .checksum = 0
+        },
+        .address = address,
+        .size = size,
+        .reserved_0 = 0,
+        .reserved_1 = 0
+    };
+
+    return send_cmd_md5(&md5_cmd, sizeof(md5_cmd), md5_out);
+}
+
+
+esp_loader_error_t loader_spi_parameters(uint32_t total_size)
+{
+    write_spi_command_t spi_cmd = {
+        .common = {
+            .direction = WRITE_DIRECTION,
+            .command = SPI_SET_PARAMS,
+            .size = 24,
+            .checksum = 0
+        },
+        .id = 0,
+        .total_size = total_size,
+        .block_size = 64 * 1024,
+        .sector_size = 4 * 1024,
+        .page_size = 0x100,
+        .status_mask = 0xFFFF,
+    };
+
+    return send_cmd(&spi_cmd, sizeof(spi_cmd), NULL);
+}
+
+
+esp_loader_error_t loader_run_stub(target_chip_t target)
+{
+    esp_loader_error_t err;
+    const esp_stub_t *stub = &esp_stub[target];
+
+    // Download segments
+    for (uint32_t seg = 0; seg < sizeof(stub->segments) / sizeof(stub->segments[0]); seg++) {
+        err = esp_loader_mem_start(stub->segments[seg].addr, stub->segments[seg].size, ESP_RAM_BLOCK);
+        if (err != ESP_LOADER_SUCCESS) {
+            return err;
+        }
+
+        size_t remain_size = stub->segments[seg].size;
+        uint8_t *data_pos = stub->segments[seg].data;
+        while (remain_size > 0) {
+            size_t data_size = MIN(ESP_RAM_BLOCK, remain_size);
+            err = esp_loader_mem_write(data_pos, data_size);
+            if (err != ESP_LOADER_SUCCESS) {
+                return err;
+            }
+            data_pos += data_size;
+            remain_size -= data_size;
+        }
+    }
+
+    err = esp_loader_mem_finish(stub->header.entrypoint);
+    if (err != ESP_LOADER_SUCCESS) {
+        return err;
+    }
+
+    // stub loader sends a custom SLIP packet of the sequence OHAI
+    uint8_t buff[4];
+    err = SLIP_receive_packet(buff, sizeof(buff) / sizeof(buff[0]));
```

Comment:
We have a problem here where the UART/USB specific protocol details are bleeding into the common source file, and we are adding an ugly `#ifdef` to a common module, defeating the purpose.

Can we move `loader_run_stub()` into `protocol_uart.c`?

Reply:
Sure.
But half of the functions inside the common module are already not really common are they? The header contains and ifdef as well.

Reply:
That is true, however
1. We can address this at a later time
2. They will be common once stub gets adapted for use over SPI (though they will depend on the stub use)

Reply:
I've moved `loader_run_stub` to `protocol_uart.c`